### DISCUSSION TITLE
Portable Power Station (PPS) support

### DISCUSSION THREAD

**Discussion by jeremywhelchel** [thomluther/anker-solix-api#95] [General]
Hey there.

Has anyone made any progress with Portable Power Station (PPS) support?
I have an Anker F3800, and my goal is to get this library (and eventually the HACS integration) to pull through some info. Particularly on discharge status.

A couple observations:
- With *just* a PPS, it seems I can't create a site in the official Anker app. So many of the API endpoints requiring `site_id` aren't going to work.
- The official app clearly shows some useful info: discharge %, temp, AC / DC input watts. Output ports used, etc
- I've hit a bunch of the endpoints listed in the current `api.py`. The only one that I see my device show up under is `bind_devices` and it looks like below:

```
{'code': 0,
 'msg': 'success!',
 'data': {'data': [{'device_sn': 'AZ...',
    'product_code': 'A1790',
    'bt_ble_id': '...',
    'bt_ble_mac': '...',
    'device_name': 'SOLIX F3800',
    'alias_name': 'SOLIX F3800',
    'img_url': 'https://public-aiot-ohi-prod.s3.dualstack.us-east-2.amazonaws.com/anker-power/public/product/2023/11/22/iot-admin/bXAgVYcpnksWkU4g/picl_A1790_normal.png',
    'link_time': ...,
    'wifi_online': True,
    'wifi_name': '',
    'relate_type': ['ble', 'wifi'],
    'charge': False,
    'bws_surplus': 0,
    'device_sw_version': 'v1.7.7',
    'has_manual': True,
    'hes_data': None}]},
 'trace_id': '...'}
```

I'd love to find more endpoints that might have useful info here.
Any thoughts on how to proceed?

I'm also very curious as to where you got the list of endpoints. If you got it via just traffic monitoring, then perhaps not having a PPS device on the account wouldn't have surfaced some of the endpoints I'm looking for.

**Comment by thomluther** [comment]
The Endpoints are extracted from the App binaries, I don't think there are much more to cover standalone devices.
What you see in the App under device details is NOT using the cloud Api, but using the MQTT servers (different infrastructure) or Bluetooth interface. I believe the MQTT servers are updated only when you use the App (pretty much overhead if there would be permanent real time traffic to the MQTT servers even if not queried by user/App).

You found the ONLY Api endpoint I saw that reports any data of standalone devices. I guess that is all you can get when the device is not configured into a power system (site).
The cloud Api only seems to receive and reflect actual values (min. 1 minute intervals) that are visible through the app in the 'home page' of the selected power system (site). Anything in the App that you see in device details is probably NOT available in the power cloud.

As an example for the Solarbank: The home page of the system does NOT show the solarbank temperature and I did not find this value in any of the endpoint data. You can see the temperature only in the Solarbank details, which is using the MQTT server interface. So neither all device data (like temperature) are sent to the power cloud even if the devices is configured into a system....

**Comment by impala454** [comment]
I recently bought a Powerhouse F2000 and would really like to implement something like this.  As @thomluther mentions this (and I assume the 3800 are Bluetooth _only_.  This was a tough sell for me over the EcoFlow but ultimately the fast switching (so I can use it as a real UPS) is what won it over.  I have done some BLE based projects in the past and searching for such a library for this device is what led me here.  I think some sleuthing of the bluetooth connection between the phone app and the device is going to be required to obtain data for these devices.  Sorry I know this is no answer for @jeremywhelchel but as long as the data is not all encrypted I think there is hope.

**Comment by thomluther** [comment]
Yes, probably, but consider following:
- BT Apis will work only locally. So when HA is too far away from the device, connection issues may occur.
- The BT interface is most likely also encrypted. If it uses EDHC encryption with assymmetric keys, it will be also difficult to get behind the used algorthms to copy the encryption mechanisms.
- The MQTT server is where all devices send their real time data. However, that obviously happens only when the app is accessing them, so this needs a trigger and I have no idea whether the duration is limited. There was already some investigation into it in #71, but I think that will be also difficult (and would also cause heavy MQTT cloud usage if all your Solix devices would report 1 sec data 24x7)

Any new interface/Api would require a separate library for usage. But if available it could potentially be integration into a common HA integration by using a different configuration entry that will utilize the different Api.
That's nothing I will spend any time on.

**Reply by doublehelix46** [reply]
I have two F3800s connected to a Anker Home Power Panel which has created a site_id. However, with the existing HACS integration, I'm not getting any values on connected Power Stations or Home Energy values. I recognize that this integration was designed for the Solarbank... but if there are any logs I could share that would be helpful to further the development - I'm more than happy to support.

**Reply by thomluther** [reply]
If you have them all configured into a power system, you can provide a system export with the [Api library tools](https://github.com/thomluther/anker-solix-api).
Please open an issue there with the attached output. The integration will not be enhanced until the Api library is able to consolidate any new device or system information.

**Reply by impala454** [reply]
@thomluther agreed, I don't think this anker-solix-api necessarily fills the bill for a bluetooth based interface, I just wanted to give @jeremywhelchel some hope that there are others out there who wish for the same thing.  It does seem like an oversight by Anker to not include a wifi interface to the device.  I am very motivated to build something that will fulfill this need, but yes would required a dedicated bluetooth computer or device to be in constant comms with the device.

**Reply by jeremywhelchel** [reply]
I'm not at home to dig in on this, but I most certainly added my F3800 to my wireless network.
Notice these lines in the json output above (I emptied the ssid for privacy)
```
    'wifi_online': True,
    'wifi_name': '',
```

So I don't believe bluetooth is required. At least for the that bind_devices interface

I would need to check if getting the other stats via the app is available outside of bluetooth range/disabled. I think so..., but maybe it's possible the api was fetching that via bluetooth (since I was sitting really close to the device).

And to Tom's point, this needs to be supported first via the API before HACS can be added. That's why I started this discussion on this repo not the hacs repo...

**Reply by jeremywhelchel** [reply]
> If you have them all configured into a power system, you can provide a system export with the [Api library tools](https://github.com/thomluther/anker-solix-api). Please open an issue there with the attached output. The integration will not be enhanced until the Api library is able to consolidate any new device or system information.

Just to be clear. This *is* the `anker-solix-api` repo. I figured a discussion was useful before making an issue.

**Reply by thomluther** [reply]
That is correct. But to upload data I believe you should open an issue here.
However, all that effort is wasted if you are not able to add the devices to a power system that you created in the Anker App.
Without power system, there is not much data accessible via the used cloud Api.
Only bind_devices will return info and that is not much and basically the few entites that you see created in HA.

**Reply by impala454** [reply]
>  I most certainly added my F3800 to my wireless network. Notice these lines in the json output above (I emptied the ssid for privacy)

Ah, so it sounds like it's just my F2000 that does not have wifi.  I'll bow out of this discussion then.

**Reply by thomluther** [reply]
Well, the device must be connected to your Wifi network, AND it must be added to a power system in the Anker App.
If both is the case, you should be able to get more data via the Cloud Api and an export of the system data should reveal the structures and fields of your devices, so they can be added potentially to the Api library.

**Reply by Guspaz** [reply]
I have a Solix F2600, which is the next generation of the F2000, and it does have a wifi connection. However, there are no options in the app related to "power systems", the app only has an option to add individual devices. I've added the F2600, which shows up as an individual device. Initially over bluetooth, and then I used the app to connect it to wifi. I'd imagine that the "power system" functionality will only be displayed if you add a device of the appropriate type.

Since I use it as a UPS, all I really want from this in Home Assistant is to get the battery percentage. Other things like power inputs/outputs are nice, but not critical. I can figure out if the mains power is online or offline many other ways, and if the mains is offline, the Anker must be on battery. So just the percentage is the key to driving other things (like automating turning things off when the battery runs low). However, I don't see that in Jeremy's data, only a "charging" boolean that wouldn't be very useful to me.

**Comment by thomluther** [comment]
For any PPS capabilities, pls review #117 for additional endpoints that were discovered.
The charging__service endpoint do not work yet on european api server, may be a matter of time.
But still then, you probably won't get any values into the api cloud if you cannot create a power system and add the device to the system.
Following main device models are supported according actual site_rules responses:
        {'rule_list': [
            {'power_site_type': 1, 'main_device_models': ['A5143'], 'device_models': ['A5143', 'A1771'], 'can_empty_site': False,
                'quantity_min_limit_map': {'A1771': 1, 'A5143': 1},'quantity_max_limit_map': {'A1771': 2, 'A5143': 1}},
            {'power_site_type': 2, 'main_device_models': ['A17C0'], 'device_models': ['A17C0', 'A5143', 'A1771'], 'can_empty_site': False,
                'quantity_min_limit_map': {'A17C0': 1}, 'quantity_max_limit_map': {'A1771': 2, 'A17C0': 2, 'A5143': 1}},
            {'power_site_type': 4, 'main_device_models': ['A17B1'], 'device_models': ['A17B1'], 'can_empty_site': True,
                'quantity_min_limit_map': None, 'quantity_max_limit_map': {'A17B1': 1}},
            {'power_site_type': 5, 'main_device_models': ['A17C1'], 'device_models': ['A17C1', 'A17X7'], 'can_empty_site': True,
                'quantity_min_limit_map': None, 'quantity_max_limit_map': {'A17C1': 1}},
            {'power_site_type': 6, 'main_device_models': ['A5341'], 'device_models': ['A5341', 'A5101', 'A5220'], 'can_empty_site': False,
                'quantity_min_limit_map': {'A5341': 1}, 'quantity_max_limit_map': {'A5341': 1}},
            {'power_site_type': 7, 'main_device_models': ['A5101'], 'device_models': ['A5101', 'A5220'], 'can_empty_site': False,
                'quantity_min_limit_map': {'A5101': 1}, 'quantity_max_limit_map': {'A5101': 6}},
            {'power_site_type': 8, 'main_device_models': ['A5102'], 'device_models': ['A5102', 'A5220'], 'can_empty_site': False,
                'quantity_min_limit_map': {'A5102': 1}, 'quantity_max_limit_map': {'A5102': 6}},
            {'power_site_type': 9, 'main_device_models': ['A5103'], 'device_models': ['A5103', 'A5220'], 'can_empty_site': False,
                'quantity_min_limit_map': {'A5103': 1}, 'quantity_max_limit_map': {'A5103': 6}}]}

Follwowing are the known main models that are required for the site:
- A5143 MI80 Inverter
- A17C0 SOLIX E1600 Solarbank
- A17B1 SOLIX Home Power Panel for SOLIX F3800
- A17C1 SOLIX E1600 Solarbank 2 Pro
- A5341 Not found yet, but eventually a X1 managing device (or installer device)
- A5101 SOLIX X1 Energy module (A5220 Battery module)
- A5102 SOLIX X1 Energy module (A5220 Battery module)
- A5103 SOLIX X1 Energy module (A5220 Battery module)
- A17C3 SOLIX E1600 Solarbank 2 Plus
